### PR TITLE
GH-54 Global require pollution

### PR DESCRIPTION
**Pull request by GordonSmith**
Signed-off-by: Gordon Smith <gordonjsmith@gmail.com>

### CODE REVIEW

File: lib/DojoAMDMainTemplatePlugin.js [hunk 1]
```diff
@@ -84,8 +84,8 @@ module.exports = class DojoAMDMainTemplatePlugin {
 				buf.push("});");
 				buf.push("");
 				buf.push("req.async = 1;");
+				buf.push("var require = req;");
```
Comment (chuckdumont) [inline lib/DojoAMDMainTemplatePlugin.js]: This won't work.  The var is not in scope for the modules.

File: lib/DojoAMDMainTemplatePlugin.js [hunk 2]
```diff
@@ -107,9 +107,9 @@ module.exports = class DojoAMDMainTemplatePlugin {
 				buf.push("\tc: createContextRequire,");
 				buf.push("\tm: dojoModuleFromWebpackModule,");
 				buf.push("\th: resolveTernaryHasExpression,");
-				buf.push("\tg: (function(){return this;})()   // Easy access to global scope");
+				buf.push("\tg: (function(){return this || window;})()   // Easy access to global scope");
```
Comment (chuckdumont) [inline lib/DojoAMDMainTemplatePlugin.js]: What value does `this` have when it is falsy?  What environment are you running in?  I have a hard time understanding how `this` would not have a value.

Comment (GordonSmith) [reply lib/DojoAMDMainTemplatePlugin.js]: Its undefined.  Which is because its ultimately inside a "use strict" section (I think)?

Comment (chuckdumont) [reply lib/DojoAMDMainTemplatePlugin.js]: What environment?

Comment (chuckdumont) [reply lib/DojoAMDMainTemplatePlugin.js]: Also, none of the bootstrap code (including the code you modified above) is placed inside a "use strict" section by webpack.  Is something else you're using wrapping the entire app within a "use strict"?

Comment (GordonSmith) [reply lib/DojoAMDMainTemplatePlugin.js]: Like I mentioned in the issue, the problem happens when this bundle is in turn bundled a second time (in rollupjs for example).  

I will try and get a sample project uploaded to GitHub by end of week.

**Comment by chuckdumont** [comment]
Closing this PR.  I've taken your changes for accessing the global scope and added the `dojo global require` mainTemplate plugin event that allows setting of global require to be customized in #94.  This will allow you to use your own plugin to achieve the results you're looking for.  Following is an example of what you want to do in your plugin.  Make sure to register the plugin **after** registering dojo-webpack-plugin.

```javascript
const ConcatSource = require("webpack-sources").ConcatSource;

module.exports = class ScopedDojoRequire {
	apply(compiler) {
		compiler.plugin("compilation", function(compilation) {
			compilation.mainTemplate.plugin("dojo global require", function(source) {
				return "";	// don't set global require
			});

			compilation.mainTemplate.plugin("require", function(source) {
				// This code, together with the 'render' plugin below, defines a scoped
				// Dojo require variable for each AMD module so that referencing 'require' from
				// within the module will yield the Dojo function.
				return source.replace(/__webpack_require__\);/g, "__webpack_require__, req);");
			});

			compilation.moduleTemplate.plugin("render", function(source, module) {
				var result = source;
				if (module.isAMD) {
					// Define a module scoped 'require' variable for AMD modules that references the
					// the Dojo require function.
					result = new ConcatSource();
					result.add("var require = arguments[3];");
					result.add(source);
				}
				return result;
			});
		});
	}
};
```

You can try out these changes using version 2.2.2-rc7 on npm.

**Pull request closed without merging**